### ISSUE TITLE
Recent docker binaries do not add PREROUTING rule

### ISSUE THREAD

**Issue by larsks** [docker/docker#10781]
With recent docker binaries (e.g., from `ee95aa1`), Docker is not creating a rule in the iptables `PREROUTING` chain.  Running `docker -D` reveals:

    DEBU[0000] /sbin/iptables, [--wait -C POSTROUTING -t nat -s 172.17.42.1/16 ! -o docker0 -j MASQUERADE] 
    DEBU[0000] /sbin/iptables, [--wait -D FORWARD -i docker0 -o docker0 -j DROP] 
    DEBU[0000] /sbin/iptables, [--wait -C FORWARD -i docker0 -o docker0 -j ACCEPT] 
    DEBU[0000] /sbin/iptables, [--wait -C FORWARD -i docker0 ! -o docker0 -j ACCEPT] 
    DEBU[0000] /sbin/iptables, [--wait -C FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -D PREROUTING -m addrtype --dst-type LOCAL -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -D OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -D PREROUTING -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -D OUTPUT -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -F DOCKER]    
    DEBU[0001] /sbin/iptables, [--wait -t nat -X DOCKER]    
    DEBU[0001] /sbin/iptables, [--wait -t nat -n -L DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -N DOCKER]    
    DEBU[0001] /sbin/iptables, [--wait -C -m addrtype --dst-type LOCAL] 
    DEBU[0001] /sbin/iptables, [--wait -C -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8] 
    DEBU[0001] /sbin/iptables, [--wait -t nat -A OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -t filter -n -L DOCKER] 
    DEBU[0001] /sbin/iptables, [--wait -C FORWARD -o docker0 -j DOCKER] 

You can see it creating an `OUTPUT` rule, but no `PREROUTING` rule.
This has several repercussions:

- Inbound traffic is being handled by the Docker proxy, rather than by
  the kernel.  This has performance implications.

- Without the `DNAT` rule, inbound traffic is hitting the `INPUT`
  chain rather than the `FORWARD` chain, which has security
  implications (e.g., by bypassing forwarding rules meant to control
  access to containerized services)

**Comment by larsks** [comment]
I am suspicious of this line:
    DEBU[0001] /sbin/iptables, [--wait -C -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8] 

It looks as if the chain name is missing from the command line.  This fails with exit status `2` instead of `1`, but I'm guessing the code does not differentiate between these two failures.

**Comment by larsks** [comment]
Ugh.  So, this is happening because we call `iptables.Exists` *before* adding `PREROUTING` to the command:

    preroute := []string{
      "-m", "addrtype",
      "--dst-type", "LOCAL"}
    if !Exists(preroute...) {

And then add `PREROUTING` in later on:

     func (c *Chain) Prerouting(action Action, args ...string) error {
	     a := []string{"-t", string(Nat), string(action), "PREROUTING"}

**Comment by larsks** [comment]
Oh, that's wrong wrong wrong.  Because the rules we're passing to `Exists` don't have `-t nat`.

**Comment by jfrazelle** [comment]
what version of iptables are you using

**Comment by jfrazelle** [comment]
So the major change from 1.4.1 to 1.5.0 with regard to iptables was https://github.com/docker/docker/pull/7003 can you describe what you are seeing?
for reference here is my output of iptables-save on master:
```
# Generated by iptables-save v1.4.21 on Fri Feb 13 16:35:49 2015
*nat
:PREROUTING ACCEPT [44434:5827410]
:INPUT ACCEPT [43928:5795824]
:OUTPUT ACCEPT [7263:460187]
:POSTROUTING ACCEPT [7263:460187]
:DOCKER - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
COMMIT
# Completed on Fri Feb 13 16:35:49 2015
# Generated by iptables-save v1.4.21 on Fri Feb 13 16:35:49 2015
*filter
:INPUT ACCEPT [255637:991137946]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [183265:42149545]
:DOCKER - [0:0]
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
COMMIT
# Completed on Fri Feb 13 16:35:49 2015
```

HTH I know how frustrating debugging iptables is.

**Comment by larsks** [comment]
This isn't a problem debugging iptables.  The problem is with the code in `iptables.go`, which is clearly not doing the right thing.  Take a walk with me...

In `NewChain`, we attempt to set up the linking rule to the `DOCKER` chain in the `nat` table with this code:

		preroute := []string{
			"-m", "addrtype",
			"--dst-type", "LOCAL"}
		if !Exists(preroute...) {
			if err := c.Prerouting(Append, preroute...); err != nil {
				return nil, fmt.Errorf("Failed to inject docker in PREROUTING chain: %s", err)
			}
		}

This call Exists with the contents of `preroute` (which does not
contain information about a table, or chain, or an action -- it's just
a simple match condition).

Exists then does:

	if _, err := Raw(append([]string{"-C"}, args...)...); err == nil {
		return true
	}

Which becomes:

    iptables -C -m addrtype --dst-type LOCAL

Which is simply invalid, because it is missing the name of a chain.  This results in:

    Bad argument `addrtype'
    Try `iptables -h' or 'iptables --help' for more information.

But even given the name of a chain, this call is *still* incorrect,
because it doesn't have a `-t <table>` argument, so it always checks
the `filter` table (whereas in this case it ought to be checking the
`nat` table).

So anyway, the call to `iptables` and always fails, but we don't catch
the failure because the code doesn't differentiate between usage
errors (errcode == 2) and other failures (errcode == 1).  Since
`iptables` fails, the code continues, and we do this:

	rule := strings.Replace(strings.Join(args, " "), "-t nat ", "", -1)
	existingRules, _ := exec.Command("iptables-save").Output()
  re := regexp.MustCompile(`[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\/[0-9]{1,2}`)

	return strings.Contains(
		re.ReplaceAllString(string(existingRules), "?"),
		re.ReplaceAllString(rule, "?"),
	)

Here, we're grabbing the entire output of `iptables-save`, and then
we're looking in it for:

    -m addrtype --dst-type LOCAL

And if that matches *anywhere*, we decide the rule exists.  In this
case, on my system, there is another rule in another table, unrelated
to Docker, that looks like this:

    -A another_chain -p tcp -m tcp --dport 8775 -m addrtype --dst-type LOCAL -j ACCEPT

Which, coincidentally, matches.  So we never create the corresponding
rule.

To summarize:

1. We're generating an invalid command line for `iptables`.  This will
   simply never work.
1. We're checking for the *match condition* of a rule against *the
   entire iptables configuration*.  This is almost guaranteed to fail.
   If we at least included the action condition (`-j DOCKER`) we would
   be better off, although we *really* need to limit the check to
   *specific* table and a *specific* chain.

**Comment by gdm85** [comment]
@larsks thanks for looking into this! @jfrazelle I am using 1.5.0 but I am not affected by this issue, as I have:
```
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
```
in my iptables.
